File: go/glyph/pool.go
package glyph

import (
	"errors"
	"fmt"
	"strings"
	"sync"
)

// Pool resolution errors
var (
	ErrPoolNotFound = errors.New("pool not found")
	ErrPoolIndex    = errors.New("pool index out of bounds")
)

// ============================================================
// Pool Reference Types
// ============================================================

// PoolRef represents a reference to a pooled value.
type PoolRef struct {
	PoolID string // Pool identifier (e.g., "S1", "O1")
	Index  int    // Index within the pool
}

// String returns the pool reference format: ^S1:0
func (p PoolRef) String() string {
	return fmt.Sprintf("^%s:%d", p.PoolID, p.Index)
}

// IsPoolRef checks if a reference string is a pool reference.
// Pool IDs start with uppercase letter followed by digit(s): S1, O1, P42
func IsPoolRef(ref string) bool {
	if len(ref) < 2 {
		return false
	}
	// Must start with uppercase letter
	if ref[0] < 'A' || ref[0] > 'Z' {
		return false
	}
	// Rest must contain at least one digit
	for i := 1; i < len(ref); i++ {
		if ref[i] >= '0' && ref[i] <= '9' {
			return true
		}
		if ref[i] < 'A' || ref[i] > 'Z' {
			return false
		}
	}
	return false
}

// ParsePoolRef parses a pool reference from "^S1:0" format.
func ParsePoolRef(input string) (*PoolRef, error) {
	if !strings.HasPrefix(input, "^") {
		return nil, fmt.Errorf("pool ref must start with ^")
	}
	input = input[1:]

	colonIdx := strings.Index(input, ":")
	if colonIdx < 0 {
		return nil, fmt.Errorf("pool ref must contain colon")
	}

	poolID := input[:colonIdx]
	if !IsPoolRef(poolID) {
		return nil, fmt.Errorf("invalid pool ID: %s", poolID)
	}

	var index int
	_, err := fmt.Sscanf(input[colonIdx+1:], "%d", &index)
	if err != nil {
		return nil, fmt.Errorf("invalid pool index: %v", err)
	}

	return &PoolRef{PoolID: poolID, Index: index}, nil
}

// ============================================================
// Pool Types
// ============================================================

// PoolKind identifies the type of pool.
type PoolKind uint8

const (
	PoolKindString PoolKind = iota // @pool.str
	PoolKindObject                 // @pool.obj
)

// Pool represents a value pool for deduplication.
type Pool struct {
	ID      string    // Pool identifier (S1, O1, etc.)
	Kind    PoolKind  // String or Object pool
	Entries []*GValue // Pool entries
}

// Get returns the value at the given index.
func (p *Pool) Get(index int) (*GValue, error) {
	if index < 0 || index >= len(p.Entries) {
		return nil, fmt.Errorf("%w: %s[%d] (len=%d)", ErrPoolIndex, p.ID, index, len(p.Entries))
	}
	return p.Entries[index], nil
}

// Add appends a value to the pool and returns its index.
func (p *Pool) Add(value *GValue) int {
	index := len(p.Entries)
	p.Entries = append(p.Entries, value)
	return index
}

// String returns the pool definition format.
func (p *Pool) String() string {
	var sb strings.Builder

	if p.Kind == PoolKindString {
		sb.WriteString("@pool.str id=")
	} else {
		sb.WriteString("@pool.obj id=")
	}
	sb.WriteString(p.ID)
	sb.WriteString(" [")

	for i, entry := range p.Entries {
		if i > 0 {
			sb.WriteString(" ")
		}
		sb.WriteString(CanonicalizeLoose(entry))
	}

	sb.WriteString("]")
	return sb.String()
}

// ============================================================
// Pool Registry
// ============================================================

// PoolRegistry manages value pools for a session.
type PoolRegistry struct {
	mu    sync.RWMutex
	pools map[string]*Pool
}

// NewPoolRegistry creates a new pool registry.
func NewPoolRegistry() *PoolRegistry {
	return &PoolRegistry{
		pools: make(map[string]*Pool),
	}
}

// Define creates or replaces a pool.
func (r *PoolRegistry) Define(pool *Pool) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.pools[pool.ID] = pool
}

// Get returns a pool by ID.
func (r *PoolRegistry) Get(id string) *Pool {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return r.pools[id]
}

// Resolve resolves a pool reference to its value.
func (r *PoolRegistry) Resolve(ref PoolRef) (*GValue, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()

	pool, ok := r.pools[ref.PoolID]
	if !ok {
		return nil, fmt.Errorf("%w: %s", ErrPoolNotFound, ref.PoolID)
	}
	return pool.Get(ref.Index)
}

// Clear removes a pool by ID.
func (r *PoolRegistry) Clear(id string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	delete(r.pools, id)
}

// ClearAll removes all pools.
func (r *PoolRegistry) ClearAll() {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.pools = make(map[string]*Pool)
}

// ============================================================
// Pool Value Type
// ============================================================

// TypePoolRef is the GType for pool references
const TypePoolRef GType = 21

// PoolRefValue creates a pool reference value.
func PoolRefValue(poolID string, index int) *GValue {
	return &GValue{
		typ:     TypePoolRef,
		poolRef: &PoolRef{PoolID: poolID, Index: index},
	}
}

// AsPoolRef returns the pool reference.
func (v *GValue) AsPoolRef() (PoolRef, error) {
	if v == nil {
		return PoolRef{}, fmt.Errorf("glyph: nil value")
	}
	if v.typ != TypePoolRef {
		return PoolRef{}, fmt.Errorf("glyph: expected poolref, got %s", v.typ)
	}
	if v.poolRef == nil {
		return PoolRef{}, nil
	}
	return *v.poolRef, nil
}

// IsPoolRef returns true if this is a pool reference.
func (v *GValue) IsPoolRef() bool {
	return v != nil && v.typ == TypePoolRef
}

// ============================================================
// Pool Emit/Parse
// ============================================================

// EmitPool emits a pool definition.
func EmitPool(pool *Pool) string {
	return pool.String()
}

// ParsePool parses a pool definition from "@pool.str id=S1 [...]".
func ParsePool(input string) (*Pool, error) {
	var kind PoolKind
	var rest string

	if strings.HasPrefix(input, "@pool.str ") {
		kind = PoolKindString
		rest = strings.TrimPrefix(input, "@pool.str ")
	} else if strings.HasPrefix(input, "@pool.obj ") {
		kind = PoolKindObject
		rest = strings.TrimPrefix(input, "@pool.obj ")
	} else {
		return nil, fmt.Errorf("expected @pool.str or @pool.obj prefix")
	}

	pool := &Pool{Kind: kind}

	// Parse id=...
	if !strings.HasPrefix(rest, "id=") {
		return nil, fmt.Errorf("expected id= after @pool")
	}
	rest = rest[3:]

	// Extract pool ID
	spaceIdx := strings.IndexAny(rest, " \t[")
	if spaceIdx < 0 {
		return nil, fmt.Errorf("missing pool entries")
	}
	pool.ID = rest[:spaceIdx]
	rest = strings.TrimLeft(rest[spaceIdx:], " \t")

	// Parse entries: [...]
	if !strings.HasPrefix(rest, "[") {
		return nil, fmt.Errorf("expected [ for pool entries")
	}

	// Find matching ]
	bracketCount := 0
	endIdx := -1
	for i := 0; i < len(rest); i++ {
		if rest[i] == '[' {
			bracketCount++
		} else if rest[i] == ']' {
			bracketCount--
			if bracketCount == 0 {
				endIdx = i
				break
			}
		}
	}
	if endIdx < 0 {
		return nil, fmt.Errorf("unclosed pool entries")
	}

	entriesStr := rest[1:endIdx]

	// Parse individual entries (simplified - space-separated values)
	entries, err := parsePoolEntries(entriesStr, kind)
	if err != nil {
		return nil, err
	}
	pool.Entries = entries

	return pool, nil
}

// parsePoolEntries parses pool entry values.
func parsePoolEntries(input string, kind PoolKind) ([]*GValue, error) {
	var entries []*GValue
	input = strings.TrimSpace(input)

	for len(input) > 0 {
		input = strings.TrimLeft(input, " \t\n")
		if input == "" {
			break
		}

		// Parse next value
		value, remaining, err := parseNextPoolValue(input)
		if err != nil {
			return nil, err
		}
		entries = append(entries, value)
		input = remaining
	}

	return entries, nil
}

// parseNextPoolValue parses the next value from pool entry string.
func parseNextPoolValue(input string) (*GValue, string, error) {
	input = strings.TrimLeft(input, " \t\n")
	if len(input) == 0 {
		return nil, "", fmt.Errorf("unexpected end of input")
	}

	// Quoted string
	if input[0] == '"' {
		endIdx := 1
		escaped := false
		for endIdx < len(input) {
			c := input[endIdx]
			if escaped {
				// Previous char was backslash, skip this char
				escaped = false
				endIdx++
				continue
			}
			if c == '\\' {
				// Next char is escaped
				escaped = true
				endIdx++
				continue
			}
			if c == '"' {
				// Found unescaped closing quote
				break
			}
			endIdx++
		}
		if endIdx >= len(input) {
			return nil, "", fmt.Errorf("unterminated quoted string")
		}
		value := unquoteBlobString(input[1:endIdx])
		return Str(value), input[endIdx+1:], nil
	}

	// Bare string or other value - find end
	endIdx := 0
	for endIdx < len(input) && input[endIdx] != ' ' && input[endIdx] != '\t' && input[endIdx] != '\n' && input[endIdx] != ']' {
		endIdx++
	}

	token := input[:endIdx]

	// Check for special values
	switch token {
	case "t", "true":
		return Bool(true), input[endIdx:], nil
	case "f", "false":
		return Bool(false), input[endIdx:], nil
	case "∅", "_", "null":
		return Null(), input[endIdx:], nil
	}

	// Check for number
	if (token[0] >= '0' && token[0] <= '9') || token[0] == '-' {
		if strings.Contains(token, ".") || strings.Contains(token, "e") || strings.Contains(token, "E") {
			var f float64
			fmt.Sscanf(token, "%f", &f)
			return Float(f), input[endIdx:], nil
		}
		var i int64
		fmt.Sscanf(token, "%d", &i)
		return Int(i), input[endIdx:], nil
	}

	// Bare string
	return Str(token), input[endIdx:], nil
}

// ============================================================
// Auto-Interning Support
// ============================================================

// AutoInternOpts configures automatic value interning.
type AutoInternOpts struct {
	MinLength   int // Minimum string length to consider (default: 50)
	MinOccurs   int // Minimum occurrences to intern (default: 2)
	MaxPoolSize int // Maximum entries per pool (default: 256)
}

// DefaultAutoInternOpts returns default auto-intern options.
func DefaultAutoInternOpts() AutoInternOpts {
	return AutoInternOpts{
		MinLength:   50,
		MinOccurs:   2,
		MaxPoolSize: 256,
	}
}

// AutoInterner tracks value occurrences and automatically interns repeated values.
// Thread-safe for concurrent use.
type AutoInterner struct {
	opts     AutoInternOpts
	registry *PoolRegistry
	mu       sync.RWMutex
	counts   map[string]int     // value -> occurrence count
	interned map[string]PoolRef // value -> pool reference
	nextPool int                // Next pool ID suffix
}

// NewAutoInterner creates a new auto-interner.
func NewAutoInterner(registry *PoolRegistry, opts AutoInternOpts) *AutoInterner {
	return &AutoInterner{
		opts:     opts,
		registry: registry,
		counts:   make(map[string]int),
		interned: make(map[string]PoolRef),
		nextPool: 1,
	}
}

// Process checks if a string should be interned and returns the appropriate value.
// Thread-safe for concurrent calls.
func (a *AutoInterner) Process(s string) *GValue {
	// Too short to consider - no locking needed
	if len(s) < a.opts.MinLength {
		return Str(s)
	}

	// Fast path: check if already interned (read lock)
	a.mu.RLock()
	if ref, ok := a.interned[s]; ok {
		a.mu.RUnlock()
		return PoolRefValue(ref.PoolID, ref.Index)
	}
	a.mu.RUnlock()

	// Slow path: need write lock
	a.mu.Lock()
	defer a.mu.Unlock()

	// Double-check after acquiring write lock
	if ref, ok := a.interned[s]; ok {
		return PoolRefValue(ref.PoolID, ref.Index)
	}

	// Count occurrence
	a.counts[s]++
	if a.counts[s] < a.opts.MinOccurs {
		return Str(s)
	}

	// Time to intern
	poolID := fmt.Sprintf("S%d", a.nextPool)
	pool := a.registry.Get(poolID)
	if pool == nil {
		pool = &Pool{ID: poolID, Kind: PoolKindString}
		a.registry.Define(pool)
	}

	if len(pool.Entries) >= a.opts.MaxPoolSize {
		// Pool full, start new one
		a.nextPool++
		poolID = fmt.Sprintf("S%d", a.nextPool)
		pool = &Pool{ID: poolID, Kind: PoolKindString}
		a.registry.Define(pool)
	}

	index := pool.Add(Str(s))
	ref := PoolRef{PoolID: poolID, Index: index}
	a.interned[s] = ref

	return PoolRefValue(ref.PoolID, ref.Index)
}

File: rust/glyph-codec/src/pool.rs
//! GLYPH Pool References
//!
//! Pool-based deduplication for strings and objects.
//! Pool refs use `^<PoolID>:<Index>` wire format; pool definitions
//! use `@pool.str id=S1 [...]` and `@pool.obj id=O1 [...]`.

use std::collections::HashMap;

use crate::error::GlyphError;
use crate::types::{GValue, MapEntry, PoolRef};

/// Pool kind: string or object.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PoolKind {
    String,
    Object,
}

impl PoolKind {
    pub fn wire_tag(&self) -> &'static str {
        match self {
            PoolKind::String => "str",
            PoolKind::Object => "obj",
        }
    }
}

/// Check if a string is a valid pool ID (uppercase letters + at least one digit).
pub fn is_pool_ref_id(s: &str) -> bool {
    if s.len() < 2 {
        return false;
    }
    let bytes = s.as_bytes();
    if !bytes[0].is_ascii_uppercase() {
        return false;
    }
    let mut saw_digit = false;
    for &b in &bytes[1..] {
        if b.is_ascii_digit() {
            saw_digit = true;
        } else if !b.is_ascii_uppercase() {
            return false;
        }
    }
    saw_digit
}

/// Parse a pool reference from `^S1:0` format.
pub fn parse_pool_ref(input: &str) -> Result<PoolRef, GlyphError> {
    if !input.starts_with('^') {
        return Err(GlyphError::Parse("pool ref must start with ^".into()));
    }
    let body = &input[1..];
    let colon = body.find(':').ok_or_else(|| GlyphError::Parse("pool ref must contain colon".into()))?;
    let pool_id = &body[..colon];
    if !is_pool_ref_id(pool_id) {
        return Err(GlyphError::Parse(format!("invalid pool ID: {}", pool_id)));
    }
    let idx_str = &body[colon + 1..];
    let index: usize = idx_str.parse().map_err(|_| GlyphError::Parse(format!("invalid pool index: {}", idx_str)))?;
    Ok(PoolRef::new(pool_id, index))
}

/// A single named pool of values.
pub struct Pool {
    pub id: String,
    pub kind: PoolKind,
    pub entries: Vec<GValue>,
}

impl Pool {
    pub fn new(id: impl Into<String>, kind: PoolKind) -> Self {
        Self {
            id: id.into(),
            kind,
            entries: Vec::new(),
        }
    }

    pub fn add(&mut self, value: GValue) -> Result<usize, GlyphError> {
        if self.kind == PoolKind::String && !value.is_str() {
            return Err(GlyphError::Parse(format!(
                "pool {} is a string pool but got non-string",
                self.id
            )));
        }
        let idx = self.entries.len();
        self.entries.push(value);
        Ok(idx)
    }

    pub fn get(&self, index: usize) -> Option<&GValue> {
        self.entries.get(index)
    }

    pub fn len(&self) -> usize {
        self.entries.len()
    }

    pub fn is_empty(&self) -> bool {
        self.entries.is_empty()
    }
}

/// Registry of pools keyed by pool ID.
pub struct PoolRegistry {
    pools: HashMap<String, Pool>,
}

impl PoolRegistry {
    pub fn new() -> Self {
        Self {
            pools: HashMap::new(),
        }
    }

    pub fn register(&mut self, pool: Pool) {
        self.pools.insert(pool.id.clone(), pool);
    }

    pub fn get(&self, pool_id: &str) -> Option<&Pool> {
        self.pools.get(pool_id)
    }

    pub fn resolve(&self, r: &PoolRef) -> Result<&GValue, GlyphError> {
        let pool = self.pools.get(&r.pool_id).ok_or_else(|| {
            GlyphError::Parse(format!("pool not found: {}", r.pool_id))
        })?;
        pool.get(r.index).ok_or_else(|| {
            GlyphError::Parse(format!(
                "pool {}[{}] out of bounds (len={})",
                r.pool_id, r.index, pool.len()
            ))
        })
    }

    pub fn ids(&self) -> Vec<String> {
        let mut ids: Vec<_> = self.pools.keys().cloned().collect();
        ids.sort();
        ids
    }
}

impl Default for PoolRegistry {
    fn default() -> Self {
        Self::new()
    }
}

/// Emit canonical pool-definition wire format.
pub fn emit_pool(pool: &Pool) -> Result<String, GlyphError> {
    use crate::loose::canonicalize_loose;

    let header = format!("@pool.{} id={}", pool.kind.wire_tag(), pool.id);
    let mut body_parts = Vec::with_capacity(pool.entries.len());
    for v in &pool.entries {
        body_parts.push(canonicalize_loose(v)?);
    }
    Ok(format!("{} [{}]", header, body_parts.join(" ")))
}

/// Recursively replace PoolRef values with their resolved pool entries.
pub fn resolve_pool_refs(value: &GValue, registry: &PoolRegistry) -> Result<GValue, GlyphError> {
    match value {
        GValue::PoolRef(r) => {
            let resolved = registry.resolve(r)?;
            Ok(resolved.clone())
        }
        GValue::List(items) => {
            let mut resolved = Vec::with_capacity(items.len());
            for v in items {
                resolved.push(resolve_pool_refs(v, registry)?);
            }
            Ok(GValue::List(resolved))
        }
        GValue::Map(entries) => {
            let mut resolved = Vec::with_capacity(entries.len());
            for e in entries {
                resolved.push(MapEntry::new(
                    e.key.clone(),
                    resolve_pool_refs(&e.value, registry)?,
                ));
            }
            Ok(GValue::Map(resolved))
        }
        GValue::Struct(s) => {
            let mut resolved = Vec::with_capacity(s.fields.len());
            for f in &s.fields {
                resolved.push(MapEntry::new(
                    f.key.clone(),
                    resolve_pool_refs(&f.value, registry)?,
                ));
            }
            Ok(GValue::struct_val(s.type_name.clone(), resolved))
        }
        GValue::Sum(s) => {
            let inner = match &s.value {
                Some(v) => Some(resolve_pool_refs(v, registry)?),
                None => None,
            };
            Ok(GValue::sum(s.tag.clone(), inner))
        }
        other => Ok(other.clone()),
    }
}
